crates/red_knot/src/symbols.rs

```diff
@@ -545,28 +546,32 @@ where
 #[derive(Debug)]
 pub(crate) struct ReachableDefinitionsIterator<'a> {
     table: &'a SymbolTable,
-    flow_node_id: FlowNodeId,
     symbol_id: SymbolId,
+    pending: Vec<FlowNodeId>,
 }
 
 impl<'a> Iterator for ReachableDefinitionsIterator<'a> {
     type Item = Definition;
 
     fn next(&mut self) -> Option<Self::Item> {
-        loop {
-            match &self.table.flow_graph.flow_nodes_by_id[self.flow_node_id] {
-                FlowNode::Start => return None,
+        while let Some(flow_node_id) = self.pending.pop() {
+            match &self.table.flow_graph.flow_nodes_by_id[flow_node_id] {
+                FlowNode::Start => return Some(Definition::None),
                 FlowNode::Definition(def_node) => {
                     if def_node.symbol_id == self.symbol_id {
-                        // we found a definition; previous definitions along this path are not
-                        // reachable
-                        self.flow_node_id = FlowGraph::start();
                         return Some(def_node.definition.clone());
                     }
-                    self.flow_node_id = def_node.predecessor;
+                    self.pending.push(def_node.predecessor);
+                }
+                FlowNode::Branch(branch_node) => {
+                    self.pending.push(branch_node.predecessor);
+                }
+                FlowNode::Phi(phi_node) => {
+                    self.pending.extend_from_slice(&phi_node.predecessors);
                 }
             }
         }
+        None
```

Comment:
Nit: You could have kept the loop

```rust

        loop {
						let flow_node_id = self.pending.pop()?;
            match &self.table.flow_graph.flow_nodes_by_id[flow_node_id] {
                FlowNode::Start => break Some(Definition::None),
                FlowNode::Definition(def_node) => {
                    if def_node.symbol_id == self.symbol_id {
                        break Some(def_node.definition.clone());
                    }
                    self.pending.push(def_node.predecessor);
                }
                FlowNode::Branch(branch_node) => {
                    self.pending.push(branch_node.predecessor);
                }
                FlowNode::Phi(phi_node) => {
                    self.pending.extend_from_slice(&phi_node.predecessors);
                }
            }
        }
```

Reply:
Good point. Do you find this clearer or more readable than the `while let`?

Reply:
Not really. I may have a slight preference to `loop` but not that it matters. I was mainly wondering if you rewrote it to a `while` loop because it wasn't clear to you how to achieve the same with a loop.

Reply:
Switched back to the loop.